api/src/verify_token.rs

```diff
@@ -0,0 +1,222 @@
+//
+// Copyright (C) 2022 WithUno, Inc.
+// All rights reserved.
+//
+// SPDX-License-Identifier: AGPL-3.0-only
+//
+
+pub use crate::store::Database;
+
+use std::result;
+
+use uno::{Mu, UnverifiedToken, VerifiedToken, VerifyMethod};
+
+use chrono::{DateTime, Utc};
+use serde_json::Error as SerdeError;
+
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum VerifyTokenError
+{
+    #[error("Serde error")]
+    Serde
+    {
+        #[from]
+        source: SerdeError,
+    },
+    #[error("Bad secret")]
+    Secret,
+    #[error("Expired")]
+    Expired,
+    #[error("Already verified")]
+    Done,
+    #[error("Not found")]
+    NotFound,
+    #[error("Unsupported schema version")]
+    Schema,
+    #[error("Unknown verify token error")]
+    Unknown,
+}
+
+type Result<T> = result::Result<T, VerifyTokenError>;
+
+pub enum PossibleToken
+{
+    Verified,
+    Unverified,
+}
+
+pub async fn get(db: &impl Database, id: &str) -> Result<PossibleToken>
+{
+    match db.get(&id).await {
+        Ok(bytes) => {
+            if serde_json::from_slice::<VerifiedToken>(&bytes).is_ok() {
+                return Ok(PossibleToken::Verified);
+            }
+
+            match serde_json::from_slice::<UnverifiedToken>(&bytes) {
+                Ok(_) => return Ok(PossibleToken::Unverified),
+                Err(e) => {
+                    return Err(VerifyTokenError::Serde { source: e });
+                },
+            }
+        },
+        Err(_) => Ok(PossibleToken::Unverified),
+    }
+}
+
+pub async fn create(
+    db: &impl Database,
+    id: &str,
+    expires_at: DateTime<Utc>,
+) -> Result<UnverifiedToken>
+{
+    if let Ok(bytes) = db.get(&id).await {
+        if serde_json::from_slice::<VerifiedToken>(&bytes).is_ok() {
+            return Err(VerifyTokenError::Done);
+        }
+
+        match serde_json::from_slice::<UnverifiedToken>(&bytes) {
+            Ok(_) => {},
+            Err(e) => {
+                return Err(VerifyTokenError::Serde { source: e });
+            },
+        };
+
+        // otherwise overwrite the unverified token, which corresponds
+        // to the case of "resend the confirmation email", etc.
```

Comment:
or a schema mismatch, I believe

Reply:
What we should do to address this concern is have the rest of the function wrapped in a schema == 0 case. I can make that change

Reply:
if we can just overwrite the old data by sending a new verification email then I think it not super valuable to try and handle old schema versions here and just let the user resend